File: clients/ios/Classes/AskAI/AskAIModels.swift
//
//  AskAIModels.swift
//  NewsBlur
//
//  Created by Claude on 2024-12-06.
//  Copyright © 2024 NewsBlur. All rights reserved.
//

import Foundation
import SwiftUI

// MARK: - AI Provider

enum AskAIProvider: String, CaseIterable, Identifiable {
    case opus = "opus"
    case gpt = "gpt-5.2"
    case gemini = "gemini-3"
    case grok = "grok-4.1"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .opus: return "Anthropic Claude Opus 4.5"
        case .gpt: return "OpenAI GPT 5.2"
        case .gemini: return "Google Gemini 3 Pro"
        case .grok: return "xAI Grok 4.1 Fast"
        }
    }

    var shortName: String {
        switch self {
        case .opus: return "Opus 4.5"
        case .gpt: return "GPT 5.2"
        case .gemini: return "Gemini 3"
        case .grok: return "Grok 4.1"
        }
    }

    var providerName: String {
        switch self {
        case .opus: return "anthropic"
        case .gpt: return "openai"
        case .gemini: return "google"
        case .grok: return "xai"
        }
    }

    var color: Color {
        switch self {
        case .opus: return Color(red: 0.85, green: 0.47, blue: 0.34) // Anthropic coral/tan
        case .gpt: return Color(red: 0.2, green: 0.65, blue: 0.45) // Green
        case .gemini: return Color(red: 0.26, green: 0.52, blue: 0.96) // Google blue
        case .grok: return Color(red: 0.1, green: 0.1, blue: 0.1) // Dark/black
        }
    }

    var textColor: Color {
        return .white
    }
}

// MARK: - Question Types

enum AskAIQuestionType: String, CaseIterable, Identifiable {
    // Summarize
    case sentence = "sentence"
    case bullets = "bullets"
    case paragraph = "paragraph"

    // Understand
    case context = "context"
    case people = "people"
    case arguments = "arguments"
    case factcheck = "factcheck"

    // Custom
    case custom = "custom"

    var id: String { rawValue }

    var displayTitle: String {
        switch self {
        case .sentence: return "Brief"
        case .bullets: return "Medium"
        case .paragraph: return "Detailed"
        case .context: return "What's the context and background?"
        case .people: return "Identify key people and relationships"
        case .arguments: return "What are the main arguments?"
        case .factcheck: return "Fact check this story"
        case .custom: return "Custom question"
        }
    }

    var subtitle: String {
        switch self {
        case .sentence: return "One sentence"
        case .bullets: return "Bullet points"
        case .paragraph: return "Full paragraph"
        default: return ""
        }
    }

    var iconName: String {
        switch self {
        case .sentence, .bullets, .paragraph: return "text.alignleft"
        case .context: return "globe"
        case .people: return "person.2"
        case .arguments: return "arrow.triangle.branch"
        case .factcheck: return "magnifyingglass"
        case .custom: return "questionmark.circle"
        }
    }

    var isSummarize: Bool {
        switch self {
        case .sentence, .bullets, .paragraph: return true
        default: return false
        }
    }

    /// Full question text shown in the response header
    var questionDescription: String {
        switch self {
        case .sentence: return "Summarize in one sentence"
        case .bullets: return "Summarize in bullet points"
        case .paragraph: return "Give a detailed summary"
        case .context: return "What's the context and background?"
        case .people: return "Identify key people and relationships"
        case .arguments: return "What are the main arguments?"
        case .factcheck: return "Fact check this story"
        case .custom: return "Custom question"
        }
    }
}

// MARK: - Conversation

struct AskAIMessage: Identifiable {
    let id = UUID()
    let role: String // "user" or "assistant"
    let content: String
}

// A completed question-response exchange
struct AskAIResponseBlock: Identifiable {
    let id = UUID()
    let questionText: String
    let model: AskAIProvider
    let responseText: String
    let isFollowUp: Bool
}

struct AskAIConversation {
    var storyHash: String
    var storyTitle: String
    var questionId: String
    var questionText: String
    var requestId: String
    var responseText: String
    var conversationHistory: [AskAIMessage]
    var completedBlocks: [AskAIResponseBlock]  // History of completed responses
    var model: AskAIProvider
    var isStreaming: Bool
    var isComplete: Bool
    var error: String?
    var usageMessage: String?

    init(storyHash: String, storyTitle: String = "") {
        self.storyHash = storyHash
        self.storyTitle = storyTitle
        self.questionId = ""
        self.questionText = ""
        self.requestId = UUID().uuidString
        self.responseText = ""
        self.conversationHistory = []
        self.completedBlocks = []
        self.model = .opus
        self.isStreaming = false
        self.isComplete = false
        self.error = nil
        self.usageMessage = nil
    }
}

// MARK: - API Response

struct AskAIQuestionResponse: Decodable {
    let code: Int
    let message: String?
    let requestId: String?
    let storyHash: String?
    let questionId: String?

    enum CodingKeys: String, CodingKey {
        case code
        case message
        case requestId = "request_id"
        case storyHash = "story_hash"
        case questionId = "question_id"
    }
}

struct AskAITranscribeResponse: Decodable {
    let code: Int
    let text: String?
    let message: String?
}

// MARK: - Socket Events

struct AskAISocketEvent {
    let type: AskAISocketEventType
    let storyHash: String
    let questionId: String
    let requestId: String
    let chunk: String?
    let error: String?
    let message: String?
}

enum AskAISocketEventType: String {
    case start = "ask_ai:start"
    case chunk = "ask_ai:chunk"
    case complete = "ask_ai:complete"
    case usage = "ask_ai:usage"
    case error = "ask_ai:error"
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/askai/AskAiModels.kt
package com.newsblur.askai

// AskAiModels.kt: Android's single place to update when AI models change.
// Raw values are stable vendor keys understood by the backend registry in
// apps/ask_ai/providers.py — only the display/short names change on upgrades.
enum class AskAiProvider(
    val rawValue: String,
    val displayName: String,
    val shortName: String,
    val providerName: String,
    val colorHex: Long,
) {
    ANTHROPIC(
        rawValue = "anthropic",
        displayName = "Anthropic Claude Sonnet 5",
        shortName = "Sonnet 5",
        providerName = "anthropic",
        colorHex = 0xFFD9735F,
    ),
    OPENAI(
        rawValue = "openai",
        displayName = "OpenAI GPT-5.6 Luna",
        shortName = "GPT-5.6 Luna",
        providerName = "openai",
        colorHex = 0xFF33A673,
    ),
    GOOGLE(
        rawValue = "google",
        displayName = "Google Gemini 3.6 Flash",
        shortName = "Gemini 3.6",
        providerName = "google",
        colorHex = 0xFF4384F5,
    ),
    XAI(
        rawValue = "xai",
        displayName = "xAI Grok 4.6",
        shortName = "Grok 4.6",
        providerName = "xai",
        colorHex = 0xFF171717,
    ),
    ;

    companion object {
        // Legacy model-specific keys ("opus", "gpt-5.2", ...) saved before
        // keys became vendor slugs (Aug 2026).
        private val legacyKeys =
            mapOf(
                "opus" to ANTHROPIC,
                "haiku" to ANTHROPIC,
                "gpt-5.2" to OPENAI,
                "gpt-5-mini" to OPENAI,
                "gemini-3" to GOOGLE,
                "gemini-flash-lite" to GOOGLE,
                "grok-4.1" to XAI,
                "grok-4.1-fast" to XAI,
            )

        fun fromRawValue(value: String?): AskAiProvider =
            entries.firstOrNull { it.rawValue == value } ?: legacyKeys[value] ?: ANTHROPIC
    }
}

enum class AskAiQuestionType(
    val rawValue: String,
    val displayTitle: String,
    val subtitle: String = "",
    val questionDescription: String,
) {
    SENTENCE(
        rawValue = "sentence",
        displayTitle = "Brief",
        subtitle = "One sentence",
        questionDescription = "Summarize in one sentence",
    ),
    BULLETS(
        rawValue = "bullets",
        displayTitle = "Medium",
        subtitle = "Bullet points",
        questionDescription = "Summarize in bullet points",
    ),
    PARAGRAPH(
        rawValue = "paragraph",
        displayTitle = "Detailed",
        subtitle = "Full paragraph",
        questionDescription = "Give a detailed summary",
    ),
    CONTEXT(
        rawValue = "context",
        displayTitle = "What's the context and background?",
        questionDescription = "What's the context and background?",
    ),
    PEOPLE(
        rawValue = "people",
        displayTitle = "Identify key people and relationships",
        questionDescription = "Identify key people and relationships",
    ),
    ARGUMENTS(
        rawValue = "arguments",
        displayTitle = "What are the main arguments?",
        questionDescription = "What are the main arguments?",
    ),
    FACTCHECK(
        rawValue = "factcheck",
        displayTitle = "Fact check this story",
        questionDescription = "Fact check this story",
    ),
    CUSTOM(
        rawValue = "custom",
        displayTitle = "Custom question",
        questionDescription = "Custom question",
    ),
    ;

    val isSummarize: Boolean
        get() = this == SENTENCE || this == BULLETS || this == PARAGRAPH
}

data class AskAiMessage(
    val role: String,
    val content: String,
)

data class AskAiResponseBlock(
    val questionText: String,
    val model: AskAiProvider,
    val responseText: String,
    val isFollowUp: Boolean,
)

data class AskAiStory(
    val storyHash: String,
    val storyTitle: String,
)

data class AskAiUiState(
    val story: AskAiStory? = null,
    val selectedModel: AskAiProvider = AskAiProvider.ANTHROPIC,
    val customQuestion: String = "",
    val completedBlocks: List<AskAiResponseBlock> = emptyList(),
    val currentQuestionId: String = "",
    val currentQuestionText: String = "",
    val currentRequestId: String = "",
    val currentResponseText: String = "",
    val isStreaming: Boolean = false,
    val isComplete: Boolean = false,
    val hasAskedQuestion: Boolean = false,
    val usageMessage: String? = null,
    val errorMessage: String? = null,
    val isRecording: Boolean = false,
    val isTranscribing: Boolean = false,
    val showAskAi: Boolean = true,
    val isArchiveTier: Boolean = false,
)
